fuel-core/src/executor.rs

```diff
@@ -897,67 +903,168 @@ impl Executor {
-        db: &Database,
-    ) -> Result<(), Error> {
-        let expected_utxo_id = |contract_id| {
-            let maybe_utxo_id = db.storage::<ContractsLatestUtxo>().get(contract_id)?;
-            let expected_utxo_id = if self.config.utxo_validation {
-                maybe_utxo_id
-                    .ok_or(Error::ContractUtxoMissing(*contract_id))?
-                    .into_owned()
-            } else {
-                maybe_utxo_id.unwrap_or_default().into_owned()
-            };
-            Result::<_, Error>::Ok(expected_utxo_id)
-        };
-
+        tx: ExecutionTypes<&mut Tx, &Tx>,
+        db: &mut Database,
+    ) -> Result<(), Error>
+    where
+        Tx: ExecutableTransaction,
+    {
         match tx {
             ExecutionTypes::Production(tx) => {
-                let inputs = match tx {
-                    Transaction::Script(script) => script.inputs_mut(),
-                    Transaction::Create(create) => create.inputs_mut(),
-                    Transaction::Mint(_) => return Ok(()),
-                };
-
-                let iter = inputs.iter_mut().filter_map(|input| match input {
-                    Input::Contract {
-                        ref mut utxo_id,
-                        ref contract_id,
-                        ..
-                    } => Some((utxo_id, contract_id)),
-                    _ => None,
-                });
-                for (utxo_id, contract_id) in iter {
-                    *utxo_id = expected_utxo_id(contract_id)?;
+                for input in tx.inputs_mut() {
+                    match input {
+                        Input::CoinSigned { tx_pointer, .. }
+                        | Input::CoinPredicate { tx_pointer, .. } => {
+                            let _ = tx_pointer;
+                            // TODO: Also calculate `tx_pointer` based on utxo's pointer.
+                            // *tx_pointer = TxPointer::new(height.into(), idx);
+                        }
+                        Input::Contract {
+                            ref mut utxo_id,
+                            ref mut balance_root,
+                            ref mut state_root,
+                            ref contract_id,
+                            ..
+                        } => {
+                            let mut contract = ContractRef::new(&mut *db, *contract_id);
+                            *utxo_id =
+                                contract.validated_utxo(self.config.utxo_validation)?;
+                            *balance_root = contract.balance_root()?;
+                            *state_root = contract.state_root()?;
+                            // TODO: Also calculate `tx_pointer` based on utxo's pointer.
+                        }
+                        _ => {}
+                    }
                 }
             }
             // Needed to convince the compiler that tx is taken by ref here
             #[allow(clippy::needless_borrow)]
```

Comment:
Does this clippy annotation still apply?

Reply:
No, we can remove